src/nbd.c

```diff
@@ -0,0 +1,1298 @@
+	g_message("%s", msg->str);
+}
+
+RaucNBDDevice *new_nbd_device(void)
+{
+	RaucNBDDevice *nbd_dev = g_malloc0(sizeof(RaucNBDDevice));
+
+	nbd_dev->sock = -1;
+
+	return nbd_dev;
+}
+
+void free_nbd_device(RaucNBDDevice *nbd_dev)
+{
+	g_return_if_fail(nbd_dev);
+
+	if (nbd_dev->index_valid) {
+		remove_nbd_device(nbd_dev, NULL);
+		/* FIXME report error */
+	}
+
+	g_free(nbd_dev);
+}
+
+RaucNBDServer *new_nbd_server(void)
+{
+	RaucNBDServer *nbd_srv = g_malloc0(sizeof(RaucNBDServer));
+
+	nbd_srv->sock = -1;
+
+	return nbd_srv;
+}
+
+void free_nbd_server(RaucNBDServer *nbd_srv)
+{
+	g_return_if_fail(nbd_srv);
+
+	if (nbd_srv->sproc) {
+		stop_nbd_server(nbd_srv, NULL);
+		/* FIXME report error */
+	}
+
+	g_free(nbd_srv->url);
+	g_free(nbd_srv->tls_cert);
+	g_free(nbd_srv->tls_key);
+	g_free(nbd_srv->tls_ca);
+	g_strfreev(nbd_srv->headers);
+	g_free(nbd_srv);
+}
+
+static int netlink_connect_cb(struct nl_msg *msg, void *arg)
+{
+	struct genlmsghdr *gnlh = nlmsg_data(nlmsg_hdr(msg));
+	RaucNBDDevice *nbd_dev = arg;
+	struct nlattr *msg_attr[NBD_ATTR_MAX + 1];
+	int ret;
+
+	ret = nla_parse(msg_attr, NBD_ATTR_MAX, genlmsg_attrdata(gnlh, 0), genlmsg_attrlen(gnlh, 0), NULL);
+	if (ret)
+		g_error("invalid response from the kernel\n");
+	if (!msg_attr[NBD_ATTR_INDEX])
+		g_error("did not receive index from the kernel\n");
+	nbd_dev->index = nla_get_u32(msg_attr[NBD_ATTR_INDEX]);
+	nbd_dev->index_valid = TRUE;
+
+	return NL_OK;
+}
+
+static struct nl_sock *netlink_connect(int *driver_id)
+{
+	struct nl_sock *nl = nl_socket_alloc();
+
+	if (!nl)
+		goto out;
+
+	if (genl_connect(nl)) { /* error */
+		nl_socket_free(nl);
+		nl = NULL;
+		goto out;
+	}
+
+	*driver_id = genl_ctrl_resolve(nl, "nbd");
+	if (*driver_id < 0) {
+		nl_close(nl);
+		nl_socket_free(nl);
+		nl = NULL;
+		goto out;
+	}
+
+out:
+	return nl;
+}
+
+gboolean setup_nbd_device(RaucNBDDevice *nbd_dev, GError **error)
+{
+	gboolean res = FALSE;
+	int driver_id;
+	struct nl_sock *nl = NULL;
+	struct nl_msg *msg = NULL;
+	struct nlattr *attr_sockets = NULL;
+	struct nlattr *attr_item = NULL;
+
+	g_return_val_if_fail(nbd_dev != NULL, FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	g_assert(nbd_dev->data_size % 4096 == 0);
+
+	g_message("configuring nbd device");
+
+	nl = netlink_connect(&driver_id);
+	if (!nl) {
+		res = FALSE;
+		g_set_error(error, G_IO_ERROR, G_IO_ERROR_FAILED, "failed to connect to netlink");
+		goto out;
+	}
+
+	msg = nlmsg_alloc();
+	if (!msg)
+		g_error("failed to allocate netlink message");
+
+	if (!genlmsg_put(msg, NL_AUTO_PORT, NL_AUTO_SEQ, driver_id, 0, 0, NBD_CMD_CONNECT, 0))
+		g_error("failed to add generic netlink headers to message");
+
+	NLA_PUT_U64(msg, NBD_ATTR_SIZE_BYTES, nbd_dev->data_size);
+	NLA_PUT_U64(msg, NBD_ATTR_BLOCK_SIZE_BYTES, 4096);
+	NLA_PUT_U64(msg, NBD_ATTR_SERVER_FLAGS, 0);
+	NLA_PUT_U64(msg, NBD_ATTR_CLIENT_FLAGS,
+			NBD_CFLAG_DESTROY_ON_DISCONNECT |
```

Comment:
The code for adding devices is there since the introduction of the netlink interface. Which error did you see?
Otherwise, perhaps leaving out `NBD_CFLAG_DESTROY_ON_DISCONNECT` has no downside?

Reply:
I've tested this again (with `./qemu-test shell` which currently uses 5.1.12):
With `NBD_CFLAG_DESTROY_ON_DISCONNECT`, the devices are used (and then removed) until none are left. RAUC still continues to work after this, as a `/dev/nbd0` is created as needed.
Without it, it keeps reusing `/dev/nbd0`.

So currently, I'd tend to leave it out, to avoid causing issues for other NBD users which are not creating devices as needed.

Reply:
Tested with v5.15.2 on hardware.

Initially, 8 nbd devices. All, including the last one show:

```
[24344.459244] block nbd7: NBD_DISCONNECT
[24344.463725] block nbd7: Disconnected due to user request.
[24344.463769] block nbd7: shutting down sockets
[24344.560521] block nbd7: NBD_DISCONNECT
[24344.560541] block nbd7: Send disconnect failed -32
```

The used devices still disappears. It also looks to me like a double-free here.

After the final nbd is gone, I get:

```
[25617.512457] nbd0: detected capacity change from 0 to 132808
[25617.513132] block nbd0: NBD_DISCONNECT
[25617.513261] block nbd0: Disconnected due to user request.
[25617.513302] block nbd0: shutting down sockets
[25617.592261] device-mapper: table: 253:0: verity: Data device lookup failed
[25617.592279] device-mapper: ioctl: error adding target to table
[25617.593687] nbd: device at index 0 is going down
```

completely preventing subsequent installations.

Reply:
The first close is triggered by `r_nbd_remove_device()` where an explicit `NBD_CMD_DISCONNECT` is sent, while the second close is triggered by the `NBD_CFLAG_DISCONNECT_ON_CLOSE`.

Leaving out both `NBD_CFLAG_DISCONNECT_ON_CLOSE` and `NBD_CFLAG_DESTROY_ON_DISCONNECT ` resolves the two above-mentioned issues on v5.15.

Reply:
> Leaving out both `NBD_CFLAG_DISCONNECT_ON_CLOSE` and `NBD_CFLAG_DESTROY_ON_DISCONNECT ` resolves the two above-mentioned issues on v5.15.

Note: this does not mean that this is already the proper solution for that issue. It's just an observation.

Reply:
I'll drop the `NBD_CFLAG_DESTROY_ON_DISCONNECT`.